lib/rouge/lexers/ghc_cmm.rb

```diff
@@ -0,0 +1,337 @@
+#
+# https://gitlab.haskell.org/ghc/ghc/wikis/commentary/compiler/cmm-syntax
+#
+module Rouge
+  module Lexers
+    class GHCCmm < RegexLexer
+      title "GHC Cmm (C--)"
+      desc "GHC Cmm is the intermediate representation of the GHC Haskell compiler"
+      tag 'ghc-cmm'
+      aliases 'cmm'
+      filenames '*.cmm', '*.dump-cmm', '*.dump-cmm-*'
+
+      macros = %r(#(?:define|include|endif|else|if))
+
+      ws = %r(\s|//.*?\n|/[*](?:[^*]|(?:[*][^/]))*[*]+/)mx
+      id = %r((?!#{macros})[\w#\$%_']+)
+
+      complex_id = %r(
+        (?:[\w#$%_']|\(\)|\(,\)|\[\]|[0-9])*
+        (?:[\w#$%_']+)
+      )mx
+
+      state :root do
+        rule %r/\s+/m, Text
+
+        # sections markers
+        rule %r/^=====.*=====$/, Generic::Heading
+
+        # timestamps
+        rule %r/^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\.\d+ UTC$/, Comment::Single
+
+        mixin :detect_section
+        mixin :preprocessor_macros
+
+        mixin :info_tbls
+        mixin :comments
+        mixin :literals
+        mixin :keywords
+        mixin :types
+        mixin :infos
+        mixin :names
+        mixin :operators
+
+        # escaped newline
+        rule %r/\\\n/, Comment::Preproc
+
+        # rest is Text
+        rule %r/./, Text
+      end
+
+      state :detect_section do
+        rule %r/(section)(\s+)/ do |m|
+          token Keyword, m[1]
+          token Text, m[2]
+          push :section
+        end
+      end
+
+      state :section do
+        rule %r/"(data|cstring|text|rodata|relrodata|bss)"/, Name::Builtin
+
+        rule %r/{/, Punctuation, :pop!
+
+        mixin :names
+        mixin :operators
+        mixin :keywords
+
+        rule %r/\s/, Text
+      end
+
+      state :preprocessor_macros do
+        rule %r/#(include|endif|else|if)/, Comment::Preproc
+
+        rule %r{
+            (\#define)
+            (#{ws}*)
+            (#{id})
+          }mx do |m|
+          token Comment::Preproc, m[1]
+          recurse m[2]
+          token Name::Label, m[3]
+        end
+      end
+
+      state :info_tbls do
+        rule %r/({ )(info_tbls)(:)/ do |m|
+          token Punctuation, m[1]
+          token Name::Entity, m[2]
+          token Punctuation, m[3]
+
+          push :info_tbls_body
+        end
+      end
+
+      state :info_tbls_body do
+        rule %r/}/, Punctuation, :pop!
+        rule %r/{/, Punctuation, :info_tbls_body
+
+        rule %r/(?=label:)/ do
+          push :label
+        end
+
+        rule %r{(\()(#{complex_id})(,)}mx do |m|
+          token Punctuation, m[1]
+          token Name::Label, m[2]
+          token Punctuation, m[3]
+        end
+
+        mixin :literals
+        mixin :infos
+        mixin :keywords
+        mixin :operators
+
+        rule %r/#{id}/, Text
+        rule %r/\s/, Text
+      end
+
+      state :label do
+        mixin :infos
+        mixin :names
+        mixin :keywords
+        mixin :operators
+
+        rule %r/[^\S\n]/, Text # Tab, space, etc. but not newline!
```

Comment:
Will something break if you were to write the pattern as `[^\S\n]+`? That would be better from a performance perspective.

Reply:
:+1: 

I've checked the other `\s` expressions, too.